File: Algorithm and Data Structure/SW Expert Academy/Problems/08673번 코딩 토너먼트1/swea8673_firestone.py
import sys
sys.stdin = open('C:\\Users\\woosa\\Desktop\\algorithms\\input.txt', 'r')

for _ in range(int(input())):
    K = int(input())
    arr = list(map(int, input().split()))
    ans = 0
    cnt = 1 << (K-1)
    while cnt >= 1:
        temp = []
        for i in range(cnt):
            ans += abs(arr[2*i]-arr[2*i+1])
            temp.append(max(arr[2*i], arr[2*i+1]))
        arr = temp
        cnt //= 2
    print("#{} {}".format(_+1, ans))

File: Algorithm and Data Structure/SW Expert Academy/Problems/08673번 코딩 토너먼트1/swea8673_firestone.cpp

#include <iostream>

using namespace std;

int abs(int a, int b) {
    if (a>=b) return a-b;
    return b-a;
}
int max(int a, int b) {
    if (a>=b) return a;
    return b;
}


void print_array(int arr[], int len) {
    for (int i=0; i<len; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main(int argc, char** argv)
{
	int test_case;
	int T;
	
	cin>>T;
	for(test_case = 1; test_case <= T; ++test_case)
	{
        int K, ans=0;
        cin >> K; // 잘 나옴
        const int a = 1<<K;
        int cnt = 1<<(K-1);
        int arr[a];
        for (int i=0; i<a; i++) {
            cin >> arr[i];
        }
        while (cnt >=1) {
            for (int i=0; i<cnt; i++) {
                ans += abs(arr[2*i], arr[2*i+1]);
                arr[i] = max(arr[2*i], arr[2*i+1]);   
            }
            cnt /=2;
        }
        printf("#%d %d\n", test_case, ans);

	}
	return 0;
}